bugsnag-android-core/src/main/java/com/bugsnag/android/SynchronizedStreamableStore.kt

```diff
@@ -0,0 +1,44 @@
+package com.bugsnag.android
+
+import android.util.JsonReader
+import java.io.File
+import java.io.InputStreamReader
+import java.util.concurrent.locks.ReentrantReadWriteLock
+import kotlin.concurrent.withLock
+
+internal class SynchronizedStreamableStore<T : JsonStream.Streamable>(
+    private val file: File
+) {
+
+    private val lock = ReentrantReadWriteLock()
+
+    fun persist(streamable: T, logger: Logger): Boolean {
+        lock.writeLock().withLock {
+            return try {
+                file.writer().use {
+                    streamable.toStream(JsonStream(it))
+                    true
+                }
+            } catch (exc: Throwable) {
+                logger.w("Failed to persist user information", exc)
+                false
+            }
+        }
+    }
+
+    fun load(logger: Logger): T? {
+        lock.readLock().withLock {
+            return try {
+                val user = file.reader().use {
+                    User.fromReader(JsonReader(it))
+                }
+                @Suppress("UNCHECKED_CAST")
+                return user as T
```

Comment:
This rather breaks the use of generics as it only reads `User` types, regardless of the type used to instantiate the store. I assume this is done with generics to provide a way to reuse for device ID and any future store - if so how would you implement a second type here?

Reply:
Yep - the plan is to reuse this class for device ID and future stores if required. I've extracted the `User.fromReader` call into a callback parameter that uses generics - this avoids the cast and supports the use of generics.